### ISSUE TITLE
hidden field/modal validation

### ISSUE THREAD

**Issue by Logan-Lim** [wet-boew/wet-boew#7898]
I'm having some issues with validation.

1. It seems that the validator is hard coded to ignore hidden fields in showErrors. When configuring wet to ignore hidden fields to validate, wet will simply prevent form submission but not actually display errors. This is causing issues with validation for collapsible menus and modals.
2. If I remove the ignore hidden from show errors the links to modal fields in the error summary are not opening the modal for that field.

Am I missing anything? Is there any solutions to these problems?

Thanks,
Logan

**Comment by kodecount** [comment]
I'm slightly lost. Do you have form inputs in a modal?

**Comment by Logan-Lim** [comment]
Yes our modals have form inputs. The fields in the modal are not actually required to save the full form. However later in our application, progress can be blocked unless these fields in this previous form are filled out. We want to allow users to navigate back to the page and show that these fields are required with a summary at the top. We still want to allow users to save the form without updating all the required fields. We are passing in a list of all fields that we detect as needed and manually trigger them through a validator showErrors. Everything in those links is working fine except modals.

We could move these modals into panels, but we would prefer to use modals if possible.

**Comment by kodecount** [comment]
If I understand you have form inputs in a modal that are apart of an form on the main page if so you can't split a form this way. WET and jQuery Validation isn't built to support it (also I can't find the previous issue that described this, but the conclusion was the same).

**Comment by kodecount** [comment]
Maybe you can try something like: https://wet-boew.github.io/v4.0-ci/demos/feedback/feedback-en.html

**Comment by Logan-Lim** [comment]
The modals and their form are all on the same page. Our application has a business rule that this document with modals must be filled out completely before the user can send it to be approved. The form is rather large so we want to show the user all the fields they are missing when they go back to update the form.

**Comment by kodecount** [comment]
_"The modals and their form are all on the same page"_ doesn't matter if the html code is on the same page. jQuery validation looks inside `<form></form>` for inputs, textareas, select, and etcetera then applies change, keyup, keydown, or etcetera type of events which validate the inputs. If their not inside the  `<form></form>` it breaks jQuery validation.

**Comment by LaurentGoderre** [comment]
Even if it work I think you would run into significant usability issues. Sounds to me like the form is too complex and should be broken down into steps.

**Comment by kodecount** [comment]
@LaurentGoderre Totally agree. Lot of sites do that. Split into multiple forms and have a progress bar (so user thinks their filling our one form even though it multiple forms split into a bunch of smaller ones).

**Comment by RobJohnston** [comment]
There was an [issue about creating a step-by-step wizard](https://github.com/wet-boew/wet-boew/issues/6367), but it went nowhere.  However, within that issue, there's a reference to[ jQuery Steps](http://www.jquery-steps.com/GettingStarted), which may be what your looking for.

**Comment by kodecount** [comment]
It a shame but would be a big undertaking. You also need to make sure you have a no script fall back (users without script would have to deal with a really long form).

**Comment by kodecount** [comment]
Also see https://github.com/wet-boew/wet-boew/issues/5457#issuecomment-44077532

**Comment by Logan-Lim** [comment]
Ya the complexity of the form is challenging. However our client requires the entire form on one page, and work is not done in a linear fashion. We have wrapped sections of our form in details elements to allow users to collapse/expand sections as needed. It seems like we'll move the modals into their own panels to solve that showErrors problem.

I've still got the issue of validation errors being displayed in the error summary for collapsed panels. I could expand all panels on submission before validation is done, but given the size of the form this doesn't seem like a great solution. Is there any work around to having hidden fields with errors display in the error summary? It seems like wet should configure the error summary to reflect the changes to the validator ignore rules.

Also as a note, we're developing this application for intranet use, so we can assume our users will have scripts enabled.

**Comment by kodecount** [comment]
Yeah between WET customization and jQuery validation moving away from hidden field validate even with the customization their many issues people experience on any platform using jQuery Validation.

I was able to get something somewhat working. You check it out here: http://kodecount.ca/demos/ . **In no way is this complete and would require some more work**

Other then the above you could experiment with jQuery Steps.

**Comment by kodecount** [comment]
Yeah between WET customization and jQuery validation moving away from hidden field validate even with the customization their many issues people experience on any platform using jQuery Validation.

Please checkout (**may require some testing**): http://kodecount.ca/demos/formvalid/

What I did was created toggle elements using `fieldset/legend` combination with a custom jQuery validation method called `data-rules-hidden` which well make `feildset/legend` visible to the DOM (to the user it seems like it still of the page (used `wb-inv` class for this (only when validating)).

Also added a `focus` event to the hidden element so when the user selects it from the error list it well make its `fieldset/legend` visible again. 

Other then the above you could experiment with jQuery Steps.

**Comment by kodecount** [comment]
Currently the method I setup has these out standing issue:
- Remove styling when JavaScript is off.
- Add wbdisable mode compatiablity.

If you want a copy of the code send me an email. I'll make sure to have a copy on my GitHub profile.

**Comment by kodecount** [comment]
Also the issue with jQuery Steps is it seem to be dead: https://github.com/rstaib/jquery-steps/issues/217

**Comment by kodecount** [comment]
I'm working on a steps version using some of the basic logic from the other example in https://github.com/wet-boew/wet-boew/issues/7898#issuecomment-280890125

**Comment by kodecount** [comment]
An example can be seen here: http://kodecount.ca/demos/formvalid/steps

**Comment by kodecount** [comment]
@duboisp @LaurentGoderre @RobJohnston 

Thoughts on https://github.com/kodecount/wet-boew/commit/136ece8bba72e6a44b4d20ad30b7ae962411628f ?

**Comment by kodecount** [comment]
So I found out the radio/checkbox button issue is not related to the plugins I'm working on. See https://github.com/wet-boew/wet-boew/issues/7903 for that issue.

**Comment by kodecount** [comment]
@Logan-Lim Let me know if any of this helps? But I think currently I would go against using modals and a form on the main page. Either split it into multiple forms or go with one of the examples I provided.

**Comment by kodecount** [comment]
@Logan-Lim Let me know if any of this helps? But I think currently I would go against using modals and a form on the main page. Either split it into multiple forms or go with one of the examples above (or something similar).

**Comment by Logan-Lim** [comment]
@kodecount Ya we've converted the modals into panels, we however aren't able to break up this page any better at the moment. Definitely going to consider using steps for future applications though. Thanks for that input. 

We are currently looking into fixing the hidden validation. In your example that you posted, it appeared to have some bugs with navigating to hidden fields. We're playing around with the validator at the moment, so we'll see if we can find a solution.

**Comment by kodecount** [comment]
Yeah using hidden inputs isn't perfect. The steps would be the simplest solution if it passes peer reviewing (which takes a long time for new plugins (since needs to be translated and accessibility tested)).
